src/pubsub.js

```diff
@@ -0,0 +1,569 @@
+              // TODO: broken https://github.com/ipfs/go-ipfs/issues/3522
+              // expect(msg).to.have.property('from', ipfs1.peerId.id)
+
+              ipfs1.pubsub.unsubscribe(topic, handler)
+
+              ipfs1.pubsub.ls((err, topics) => {
+                expect(err).to.not.exist
+                expect(topics).to.be.empty
+                check()
+              })
+            }
+
+            ipfs1.pubsub.subscribe(topic, handler, (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.publish(topic, 'hi', check)
+            })
+          })
+
+          it('attaches multiple event listeners', (done) => {
+            const check = makeCheck(3, done)
+            const handler1 = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hello')
+
+              ipfs1.pubsub.unsubscribe(topic, handler1)
+
+              series([
+                (cb) => ipfs1.pubsub.ls(cb),
+                (cb) => {
+                  ipfs1.pubsub.unsubscribe(topic, handler2)
+                  cb()
+                },
+                (cb) => ipfs1.pubsub.ls(cb)
+              ], (err, res) => {
+                expect(err).to.not.exist
+
+                // Still subscribed as there is one listener left
+                expect(res[0]).to.be.eql([topic])
+                // Now all listeners are gone no subscription anymore
+                expect(res[2]).to.be.eql([])
+                check()
+              })
+            }
+
+            const handler2 = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hello')
+              check()
+            }
+
+            parallel([
+              (cb) => ipfs1.pubsub.subscribe(topic, handler1, cb),
+              (cb) => ipfs1.pubsub.subscribe(topic, handler2, cb)
+            ], (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.publish(topic, 'hello', check)
+            })
+          })
+
+          it('discover options', (done) => {
+            const check = makeCheck(2, done)
+
+            const handler = (msg) => {
+              expect(msg.data.toString()).to.be.eql('hi')
+              ipfs1.pubsub.unsubscribe(topic, handler)
+              check()
+            }
+
+            ipfs1.pubsub.subscribe(topic, {
+              discover: true
+            }, handler, (err) => {
+              expect(err).to.not.exist
+              ipfs1.pubsub.publish(topic, 'hi', check)
+            })
+          })
+        })
+      })
+
+      describe('multiple nodes connected', () => {
+        before((done) => {
+          parallel([
+            (cb) => ipfs1.swarm.connect(ipfs2.peerId.addresses[0], cb),
+            (cb) => ipfs2.swarm.connect(ipfs3.peerId.addresses[0], cb),
+            (cb) => ipfs1.swarm.connect(ipfs3.peerId.addresses[0], cb)
+          ], done)
+        })
+
+        describe('.peers', () => {
+          it('returns an error when not subscribed to a topic', (done) => {
```

Comment:
There was an open question in this interface. I personally don't understand why the question "Tell me who I know that is subscribed to a topic" needs to fail in case I'm not interested to the topic.

@haadcode could you clarify your argument for it to be this way? 

@gavinmcdermott @dignifiedquire thoughts?